pkg/schedule/schedulers/balance_range.go

```diff
@@ -70,13 +77,59 @@ type balanceRangeSchedulerConfig struct {
 type balanceRangeSchedulerJob struct {
 	JobID   uint64          `json:"job-id"`
 	Role    Role            `json:"role"`
-	Engine  string          `json:"engine"`
+	Engine  engine          `json:"engine"`
 	Timeout time.Duration   `json:"timeout"`
 	Ranges  []core.KeyRange `json:"ranges"`
 	Alias   string          `json:"alias"`
+	Start   *time.Time      `json:"start,omitempty"`
+	Finish  *time.Time      `json:"finish,omitempty"`
+	Create  time.Time       `json:"create"`
 	Status  JobStatus       `json:"status"`
 }
 
+func (conf *balanceRangeSchedulerConfig) begin(job *balanceRangeSchedulerJob) bool {
+	conf.Lock()
+	defer conf.Unlock()
+	if job.Status != pending {
+		return false
+	}
+	now := time.Now()
+	job.Start = &now
+	job.Status = running
+	if err := conf.save(); err != nil {
+		log.Warn("failed to persist config", zap.Error(err), zap.Uint64("job-id", job.JobID))
+		return false
+	}
+	return true
+}
+
+func (conf *balanceRangeSchedulerConfig) finish(job *balanceRangeSchedulerJob) bool {
+	conf.Lock()
+	defer conf.Unlock()
+	if job.Status != running {
+		return false
+	}
+	now := time.Now()
+	job.Finish = &now
+	if err := conf.save(); err != nil {
+		log.Warn("failed to persist config", zap.Error(err), zap.Uint64("job-id", job.JobID))
+		return false
+	}
+	return true
+}
+
+func (conf *balanceRangeSchedulerConfig) peek() *balanceRangeSchedulerJob {
+	conf.RLock()
+	defer conf.RLock()
+	for _, job := range conf.jobs {
+		if job.Status == finished {
```

Comment:
Where do we change the status to finished?

Reply:
good catch

Reply:
If all ranges is balanced, we still need to wait for timeout?

Reply:
yes in this pr. I will add other conditions to check whether the distribution is balanced in next PR.